### PR TITLE
Add Docker containers for the app.

### PR DESCRIPTION
**Pull request by willrlin**
# Pull Request Template

## Issue

Addresses CityOfDetroit/affordable-housing-app#93
- [x] This change addresses the issue in full
- [ ] This change addresses only certain aspects of the issue
- [ ] This change is a dependency for another issue
- [ ] This change has a dependency from another issue

## Description

In order to easily deploy the website to any popular service, we needed to containerize it. This adds Dockerfile's for the 4 major portions of the site:
- ui-components
- sites/public
- sites/partners
- backend/core

It also adds a `docker-compose.yml` file for easy starting of development:
`docker-compose up`

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Prototype/POC (not to merge)
- [ ] This change is a refactor/address technical debt
- [ ] This change requires a documentation update
- [ ] This change requires a SQL Script

## How Can This Be Tested/Reviewed?

Can be verified by running `docker-compose up` and viewing the server.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have reviewed the changes in a desktop view
- [x] I have reviewed the changes in a mobile view
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] I have assigned reviewers
- [ ] I have updated the changelog to include a description of my changes

### CODE REVIEW

**Comment by willrlin** [comment]
The backend-core build is failing because a migration needs to be run on the cloud deployment. For development, the docker-compose file runs the migration prior to launching backend/core.

**Comment by abbiefarr** [comment]
Could you help me understand the build issue a bit more? Shouldn't the auto deployment of Dockerfile.run_migration have fixed it?

**Comment by willrlin** [comment]
I chose not to auto-deploy the Dockerfile.run_migration version because it takes a long time to start up since it's running the migration when it starts the container, which we'd only need every once in a while when the schema changes. But if you tried to spin up more containers, it would try to run those for each container it spins up. I'm trying to get a workflow going where we can run the migration locally, which alleviates the need for doing it inside the container.

File: Dockerfile.sites-partners [hunk 1]
```diff
@@ -0,0 +1,64 @@
+FROM node:12.16-alpine AS development
+# Check https://github.com/nodejs/docker-node/tree/b4117f9333da4138b03a546ec926ef50a31506c3#nodealpine to understand why libc6-compat might be needed.
+RUN apk add --no-cache libc6-compat
+
+WORKDIR /usr/src/app/
+
+COPY . ./
```
Comment (avaleske) [inline Dockerfile.sites-partners:7]: Can you add a comment about what's happening here.

File: backend/core/src/shared/shared.module.ts [hunk 2]
```diff
@@ -12,7 +12,7 @@ import Joi from "joi"
           .default("development"),
         EMAIL_API_KEY: Joi.string().required(),
         EMAIL_FROM_ADDRESS: Joi.string().required(),
-        DATABASE_URL: Joi.string().required(),
+        DATABASE_URL: Joi.string().optional(),
```
Comment (avaleske) [inline backend/core/src/shared/shared.module.ts:15]: Was this a temporary change to get things working, or do we want it to be optional?

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+      - ./sites/public:/usr/src/app/sites/public
+      - /usr/src/app/node_modules
+      - /usr/src/app/sites/public/node_modules
+    ports:
+      # TODO: configure via .env separate from ./sites/public/.env
+      - "3000:3000"
+    env_file:
+      - ./sites/public/.env
+    environment:
+      # TODO: login currently doesn't work, likely needs a separate env variable to
```
Comment (avaleske) [inline docker-compose.yml:20]: Can you make an issue for this?

**Review by avaleske — APPROVED** [review]

File: backend/core/src/shared/shared.module.ts (hunk 2 above)
Comment (willrlin) [reply backend/core/src/shared/shared.module.ts:15]: Temporary, I have a local commit to revert that part. I didn't want to keep kicking off builds for every comment I address.

File: docker-compose.yml (hunk 3 above)
Comment (willrlin) [reply docker-compose.yml:20]: That's CityOfDetroit/affordable-housing-app#84. I'll add that number in the comment.

File: Dockerfile.sites-partners (hunk 1 above)
Comment (willrlin) [reply Dockerfile.sites-partners:8]: Done!

**Pull request merged**